core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -0,0 +1,614 @@
+        s.success = true
+        s.driverState = "Driver is queued for launch"
+      } else if (launchedDrivers.contains(submissionId)) {
+        s.success = true
+        s.driverState = "Driver is running"
+        launchedDrivers(submissionId).taskState.foreach(state => s.message = state.toString)
+      } else if (finishedDrivers.exists(s => s.submission.submissionId.equals(submissionId))) {
+        s.success = true
+        s.driverState = "Driver already finished"
+        finishedDrivers.find(d => d.submission.submissionId.equals(submissionId)).get.taskState
+          .foreach(state => s.message = state.toString)
+      } else if (pendingRetryDrivers.exists(_.submissionId.equals(submissionId))) {
+        val status = pendingRetryDrivers.find(_.submissionId.equals(submissionId))
+          .get.retryState.get.lastFailureStatus
+        s.success = true
+        s.driverState = "Driver failed and retrying"
+        s.message = status.toString
+      } else {
+        s.success = false
+        s.driverState = "Driver not found"
+      }
+    }
+    s
+  }
+
+  override def offerRescinded(driver: SchedulerDriver, offerId: OfferID): Unit = {}
+  override def disconnected(driver: SchedulerDriver): Unit = {}
+  override def reregistered(driver: SchedulerDriver, masterInfo: MasterInfo): Unit = {
+    logInfo(s"Framework re-registered with master ${masterInfo.getId}")
+  }
+  override def slaveLost(driver: SchedulerDriver, slaveId: SlaveID): Unit = {}
+  override def error(driver: SchedulerDriver, error: String): Unit = {}
+
+  /**
+   * Check if the task state is a recoverable state that we can relaunch the task.
+   * Task state like TASK_ERROR are not relaunchable state since it wasn't able
+   * to be validated by Mesos.
+   */
+  def shouldRelaunch(state: MesosTaskState): Boolean = {
+    state == MesosTaskState.TASK_FAILED ||
+      state == MesosTaskState.TASK_KILLED ||
+      state == MesosTaskState.TASK_LOST
+  }
+
+  override def statusUpdate(driver: SchedulerDriver, status: TaskStatus): Unit = {
+    val taskId = status.getTaskId.getValue
+    stateLock.synchronized {
+      if (launchedDrivers.contains(taskId)) {
+        if (status.getReason == Reason.REASON_RECONCILIATION &&
+          !pendingRecover.contains(taskId)) {
+          // Task has already received update and no longer requires reconciliation.
+          return
+        }
+        val state = launchedDrivers(taskId)
+        // Check if the driver is supervise enabled and can be relaunched.
+        if (state.submission.supervise && shouldRelaunch(status.getState)) {
+          removeFromLaunchedDrivers(taskId)
+          val retryState: Option[RetryState] = state.submission.retryState
+          val (retries, waitTimeSec) = if (retryState.isDefined) {
+            (retryState.get.retries + 1,
+              Math.min(maxRetryWaitTime, retryState.get.waitTime * 2))
+          } else {
+            (1, 1)
+          }
```

Comment:
minor: This can be rewritten in a more scala friendly way:
```
val (retries, waitTimeSec) = state.submission.retryState
  .map { rs => (rs.retries + 1, Math.min(maxRetryWaitTime, rs.waitTime * 2) }
  .getOrElse { (1, 1) }
```